Sources/BraintreeCoreSwift/BTAPIClientSwift.swift

```diff
@@ -1,5 +1,441 @@
+
+        return components.url
+    }
+
+    static func authorizationType(forAuthorization authorization: String) -> BTAPIClientAuthorization? {
+        let pattern: String = "([a-zA-Z0-9]+)_[a-zA-Z0-9]+_([a-zA-Z0-9_]+)"
+        guard let regularExpression = try? NSRegularExpression(pattern: pattern) else { return nil }
+
+        let tokenizationKeyMatch: NSTextCheckingResult? = regularExpression.firstMatch(in: authorization, options: [], range: NSRange(location: 0, length: authorization.count))
+
+        return tokenizationKeyMatch != nil ? .tokenizationKey : .clientToken
+    }
+
+    static func scheme(forEnvironment environment: String) -> String {
+        environment.lowercased() == "development" ? "http" : "https"
+    }
+
+    static func host(forEnvironment environment: String, httpType: BTAPIClientHTTPTypeSwift) -> String? {
+        var host: String? = nil
+        let environmentLowercased: String = environment.lowercased()
+
+        switch httpType {
+        case .gateway:
+            if environmentLowercased == "sandbox" {
+                host = "api.sandbox.braintreegateway.com"
+            } else if environmentLowercased == "production" {
+                host = "api.braintreegateway.com:443"
+            } else if environmentLowercased == "development" {
+                host = "localhost:3000"
+            }
+
+        case .graphQLAPI:
+            if environmentLowercased == "sandbox" {
+                host = "payments.sandbox.braintree-api.com"
+            } else if environmentLowercased == "development" {
+                host = "localhost:8080"
+            } else {
+                host = "payments.braintree-api.com"
+            }
+
+        default:
+            host = nil
+        }
+
+        return host
+    }
+
+    static func clientApiBasePath(forMerchantID merchantID: String) -> String {
+        "/merchants/\(merchantID)/client_api"
+    }
+
+    // MARK: - Internal Methods
+
+    func graphQLURL(forEnvironment environment: String) -> URL? {
+        var components: URLComponents = URLComponents()
+        components.scheme = Self.scheme(forEnvironment: environment)
+
+        guard let host: String = Self.host(forEnvironment: environment, httpType: .graphQLAPI) else { return nil }
+        let hostComponents: [String] = host.components(separatedBy: ":")
+
+        if hostComponents.count == 0 {
+            return nil
+        }
+
+        components.host = hostComponents.first
+
+        if hostComponents.count > 1 {
+            let portString: String = hostComponents[1]
+            components.port = Int(portString)
+        }
+
+        components.path = "/graphql"
+        return components.url
+    }
+
+    func httpForType(_ httpType: BTAPIClientHTTPTypeSwift) -> BTHTTP? {
+        switch httpType {
+        case .gateway:
+            return http
+        case .braintreeAPI:
+            return apiHTTP
+        case .graphQLAPI:
+            return graphQLHTTP
+        }
+    }
```

Comment:
Is this reasonable? The optional return type leads to optional chaining in the `post` and `get` requests which would be nice to avoid. we don't have to repeat the line `return http` as I suggest below, I just wanted to make sure the `gateway` case was explicitly decleared, and figured that `http` makes the most sense as a default.

I also moved the word `for` into the params just as a nitpick. 

```suggestion
    func http(for _ httpType: BTAPIClientHTTPTypeSwift) -> BTHTTP {
           switch httpType {
         case .gateway:
             return http
         case .braintreeAPI:
             return apiHTTP
         case .graphQLAPI:
             return graphQLHTTP
         default:
             return .http
         }
        }
    }
```

Reply:
Actually... could this be a computed variable on the `BTAPIClientHTTPTypeSwift` enum itself?

Reply:
🤔  so this is a bit tricky because here we are actually returning the property defined here:
```
    var http: BTHTTP?
    var apiHTTP: BTAPIHTTP?
    var graphQLHTTP: BTGraphQLHTTP?
```

I don't think we want the enum to have access to the properties defined here. Making this required would require us to nil coalesce instead of provide a default and I don't think we want to provide a property when the properties defined at the top are nil, right?

Reply:
ooooh I read this as returning enum instances